src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    ///
+    /// Returns `(k, offset)`, where `k` is layout of the concatenated
+    /// record and `offset` is the relative location, in bytes, of the
+    /// start of the `next` embedded witnin the concatenated record
+    /// (assuming that the record itself starts at offset 0).
+    ///
+    /// (The `offset` is always the same as `self.size()`; we use this
+    ///  signature out of convenience in matching the signature of
+    ///  `fn extend`.)
+    ///
+    /// Requires no arithmetic overflow from inputs.
+    /// (See also the `fn extend_packed` checked variant.)
+    pub unsafe fn extend_packed_unchecked(&self, next: Self) -> (Self, usize) {
+        self.extend_packed(next).unwrap()
+    }
+
+    /// Creates a layout describing the record for a `[T; n]`.
+    ///
+    /// On zero `n`, zero-sized `T`, or arithmetic overflow, returns `None`.
+    pub fn array<T>(n: usize) -> Option<Self> {
+        Layout::new::<T>()
+            .repeat(n)
+            .map(|(k, offs)| {
+                debug_assert!(offs == mem::size_of::<T>());
+                k
+            })
+    }
+
+    /// Creates a layout describing the record for a `[T; n]`.
+    ///
+    /// Requires nonzero `n`, nonzero-sized `T`, and no arithmetic
+    /// overflow; otherwise behavior undefined.
+    pub fn array_unchecked<T>(n: usize) -> Self {
+        Layout::array::<T>(n).unwrap()
+    }
+
+}
+
+/// The `AllocErr` error specifies whether an allocation failure is
+/// specifically due to resource exhaustion or if it is due to
+/// something wrong when combining the given input arguments with this
+/// allocator.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum AllocErr {
+    /// Error due to hitting some resource limit or otherwise running
+    /// out of memory. This condition strongly implies that *some*
+    /// series of deallocations would allow a subsequent reissuing of
+    /// the original allocation request to succeed.
+    Exhausted { request: Layout },
+
+    /// Error due to allocator being fundamentally incapable of
+    /// satisfying the original request. This condition implies that
+    /// such an allocation request will never succeed on the given
+    /// allocator, regardless of environment, memory pressure, or
+    /// other contextual condtions.
+    ///
+    /// For example, an allocator that does not support zero-sized
+    /// blocks can return this error variant.
+    Unsupported { details: &'static str },
+}
+
+impl AllocErr {
+    pub fn invalid_input(details: &'static str) -> Self {
+        AllocErr::Unsupported { details: details }
+    }
+    pub fn is_memory_exhausted(&self) -> bool {
+        if let AllocErr::Exhausted { .. } = *self { true } else { false }
+    }
+    pub fn is_request_unsupported(&self) -> bool {
+        if let AllocErr::Unsupported { .. } = *self { true } else { false }
+    }
+}
+
+/// The `CannotReallocInPlace` error is used when `fn realloc_in_place`
```

Comment:
This reflects a personal preference on my part but I'm happy to go along whatever the official doc format is.

I did see [discussion](https://github.com/rust-lang/rfcs/pull/1946#issuecomment-284820064) recently that our docs do have a problem in that we don't have a standard way to differentiate between the three Rust namespaces when writing links; IMO a prefix like `fn` would seem like a reasonable way to explicitly disambiguate, but I'm probably in the minority in my opinion about this...